File: a2a-core/src/main/java/com/fox/a2a/core/A2AAgent.java
package com.fox.a2a.core;

import com.fox.a2a.core.handler.MessageHandler;
import com.fox.a2a.core.handler.TaskHandler;
import com.fox.a2a.core.pubsub.PubSubSubscriber;
import com.fox.a2a.proto.AgentInfo;
import com.fox.a2a.proto.Message;
import lombok.extern.slf4j.Slf4j;

import java.io.Closeable;
import java.util.Arrays;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;

/**
 * A2A Agent 门面类
 * 整合 A2AClient（对外通信）和 AgentServer（接收消息），提供统一的 Agent 操作入口
 * 支持链式调用配置消息处理器、任务处理器和 topic 订阅
 */
@Slf4j
public class A2AAgent implements Closeable {

    /** Agent 配置 */
    private final A2AConfig config;

    /** 核心客户端（负责对外通信） */
    private final A2AClient client;

    /** gRPC 服务器（负责接收消息） */
    private final AgentServer server;

    /** PubSub 订阅者 */
    private PubSubSubscriber pubSubSubscriber;

    /** 待订阅的 topic 列表（在 start() 前收集） */
    private final List<String> pendingTopics = new ArrayList<>();

    /**
     * 私有构造函数，通过 create() 工厂方法创建实例
     *
     * @param config A2A 配置
     * @param client 核心客户端
     * @param server gRPC 服务器
     */
    private A2AAgent(A2AConfig config, A2AClient client, AgentServer server) {
        this.config = config;
        this.client = client;
        this.server = server;
    }

    /**
     * 工厂方法：根据配置创建 A2AAgent 实例
     *
     * @param config A2A 配置
     * @return A2AAgent 实例
     */
    public static A2AAgent create(A2AConfig config) {
        // 创建客户端
        A2AClient client = new A2AClient(config);
        // 创建服务器（handler 后续通过 onMessage/onTask 设置）
        AgentServer server = new AgentServer(config.getAgentPort());
        return new A2AAgent(config, client, server);
    }

    /**
     * 启动 Agent
     * 1. 启动 gRPC 服务器（开始监听端口）
     * 2. 如果配置了 autoRegister，向 Registry 注册
     * 3. 如果有待订阅的 topic，启动 PubSub 订阅
     *
     * @throws Exception 启动失败时抛出
     */
    public void start() throws Exception {
        log.info("正在启动 A2AAgent [{}]...", config.getAgentId());

        // 启动 gRPC 服务器
        server.start();
        log.info("AgentServer 已启动，端口: {}", config.getAgentPort());

        // 自动注册到 Registry
        if (config.isAutoRegister()) {
            client.register();
        }

        // 启动 PubSub 订阅（如果有待订阅的 topic）
        if (!pendingTopics.isEmpty()) {
            startPubSubSubscription(pendingTopics);
        }

        log.info("A2AAgent [{}] 启动完成", config.getAgentId());
    }

    /**
     * 设置消息处理器（链式调用）
     * 必须在 start() 之前调用
     *
     * @param handler 消息处理器实现
     * @return this（支持链式调用）
     */
    public A2AAgent onMessage(MessageHandler handler) {
        server.setMessageHandler(handler);
        log.debug("已设置 MessageHandler: {}", handler.getClass().getSimpleName());
        return this;
    }

    /**
     * 设置任务处理器（链式调用）
     * 必须在 start() 之前调用
     *
     * @param handler 任务处理器实现
     * @return this（支持链式调用）
     */
    public A2AAgent onTask(TaskHandler handler) {
        server.setTaskHandler(handler);
        log.debug("已设置 TaskHandler: {}", handler.getClass().getSimpleName());
        return this;
    }

    /**
     * 订阅指定 topic（链式调用）
     * 如果在 start() 之前调用，topic 会在 start() 时统一订阅
     * 如果在 start() 之后调用，立即建立订阅
     *
     * @param topics 要订阅的 topic 名称（可变参数）
     * @return this（支持链式调用）
     */
    public A2AAgent subscribe(String... topics) {
        List<String> topicList = Arrays.asList(topics);
        if (pubSubSubscriber != null && pubSubSubscriber.isRunning()) {
            // 已启动，直接追加订阅（重新订阅包含新 topic）
            pendingTopics.addAll(topicList);
            startPubSubSubscription(pendingTopics);
        } else {
            // 未启动，加入待订阅列表
            pendingTopics.addAll(topicList);
        }
        log.debug("已添加订阅 topics: {}", topicList);
        return this;
    }

    /**
     * 启动 PubSub 订阅
     *
     * @param topics 要订阅的 topic 列表
     */
    private void startPubSubSubscription(List<String> topics) {
        // 如果已有订阅，先取消
        if (pubSubSubscriber != null && pubSubSubscriber.isRunning()) {
            pubSubSubscriber.unsubscribe();
        }
        pubSubSubscriber = new PubSubSubscriber(
                client.getPubSubAsyncStub(),
                config.getAgentId(),
                client.getSessionToken()
        );
        pubSubSubscriber.subscribe(topics, message -> {
            log.debug("收到 PubSub 消息，topic: {}", message.getTopic());
            // PubSub 消息通过 MessageHandler 的 handleMessage 分发
        });
    }

    // ==================== 代理 A2AClient 的方法 ====================

    /**
     * 按类型发现 Agent
     *
     * @param agentType Agent 类型
     * @return 匹配的 AgentInfo 列表
     */
    public List<AgentInfo> discover(String agentType) {
        return client.discover(agentType);
    }

    /**
     * 按能力发现 Agent
     *
     * @param capability 能力名称
     * @return 匹配的 AgentInfo 列表
     */
    public List<AgentInfo> discoverByCapability(String capability) {
        return client.discoverByCapability(capability);
    }

    /**
     * 同步请求/响应
     *
     * @param toAgentId 目标 Agent ID
     * @param payload   请求载荷
     * @return 响应消息
     */
    public Message sendRequest(String toAgentId, Map<String, Object> payload) {
        return client.sendRequest(toAgentId, payload);
    }

    /**
     * 单向发送消息
     *
     * @param toAgentId 目标 Agent ID
     * @param topic     消息主题
     * @param payload   消息载荷
     * @return 消息 ID
     */
    public String send(String toAgentId, String topic, Map<String, Object> payload) {
        return client.send(toAgentId, topic, payload);
    }

    /**
     * 发布消息到 topic
     *
     * @param topic   主题名称
     * @param payload 消息载荷
     * @return 消息 ID
     */
    public String publish(String topic, Map<String, Object> payload) {
        return client.publish(topic, payload);
    }

    /**
     * 委托任务给目标 Agent
     *
     * @param toAgentId 目标 Agent ID
     * @param taskType  任务类型
     * @param input     任务输入
     * @return 任务 ID
     */
    public String delegateTask(String toAgentId, String taskType, Map<String, Object> input) {
        return client.delegateTask(toAgentId, taskType, input);
    }

    /**
     * 查询任务状态
     *
     * @param taskId 任务 ID
     * @return 任务信息
     */
    public com.fox.a2a.proto.Task getTaskStatus(String taskId) {
        return client.getTaskStatus(taskId);
    }

    /**
     * 取消任务
     *
     * @param taskId 任务 ID
     * @param reason 取消原因
     * @return 是否取消成功
     */
    public boolean cancelTask(String taskId, String reason) {
        return client.cancelTask(taskId, reason);
    }

    /**
     * 获取底层 A2AClient（高级用法）
     *
     * @return A2AClient
     */
    public A2AClient getClient() {
        return client;
    }

    /**
     * 获取底层 AgentServer（高级用法）
     *
     * @return AgentServer
     */
    public AgentServer getServer() {
        return server;
    }

    /**
     * 获取配置
     *
     * @return A2AConfig
     */
    public A2AConfig getConfig() {
        return config;
    }

    /**
     * 阻塞等待服务器终止（用于主线程保活）
     *
     * @throws InterruptedException 线程中断时抛出
     */
    public void awaitTermination() throws InterruptedException {
        server.awaitTermination();
    }

    /**
     * 关闭 Agent
     * 停止 PubSub 订阅、关闭客户端、停止服务器
     */
    @Override
    public void close() {
        log.info("正在关闭 A2AAgent [{}]...", config.getAgentId());
        // 停止 PubSub 订阅
        if (pubSubSubscriber != null && pubSubSubscriber.isRunning()) {
            pubSubSubscriber.unsubscribe();
        }
        // 关闭客户端（包含注销和停止心跳）
        client.close();
        // 停止 gRPC 服务器
        server.stop();
        log.info("A2AAgent [{}] 已关闭", config.getAgentId());
    }
}

File: a2a-node-sdk/src/a2a-agent.ts
/**
 * A2A Framework Node.js SDK - A2AAgent
 * 门面类，整合 A2AClient（客户端）和 AgentServer（服务端）
 * 提供统一的 Agent 生命周期管理和消息处理接口
 */

import {
  A2AConfig,
  AgentInfo,
  Task,
  MessageHandler,
  RequestHandler,
  TaskHandler,
  PubSubHandler,
} from './types';
import { A2AClient } from './a2a-client';
import { AgentServer } from './a2a-server';
import { PubSubSubscriber } from './pubsub-subscriber';
import { formatError } from './utils';

// ==================== A2AAgent 类 ====================

/**
 * A2A Agent 门面类
 * 整合客户端（服务发现、消息发送）和服务端（接收消息）功能
 * 提供简洁的 API 用于构建 A2A 应用
 *
 * @example
 * ```typescript
 * const agent = A2AAgent.create({
 *   agentId: 'my-agent-001',
 *   agentName: 'My Agent',
 *   agentType: 'llm-agent',
 *   agentPort: 8080,
 *   registryHost: 'localhost',
 *   registryPort: 9090,
 *   jwtToken: 'your-jwt-token',
 * });
 *
 * agent.onRequest(async (fromAgentId, messageId, payload) => {
 *   return { result: 'ok' };
 * });
 *
 * await agent.start();
 * ```
 */
export class A2AAgent {
  /** A2A 客户端（负责对外通信） */
  private client: A2AClient;
  /** Agent gRPC 服务端（负责接收消息） */
  private server: AgentServer;
  /** PubSub 订阅者列表 */
  private subscribers: PubSubSubscriber[] = [];
  /** 配置 */
  private config: A2AConfig;
  /** 是否已启动 */
  private started: boolean = false;

  /**
   * 私有构造函数，请使用 A2AAgent.create() 创建实例
   */
  private constructor(config: A2AConfig) {
    this.config = config;

    // 创建客户端
    this.client = new A2AClient(config);

    // 创建服务端（监听指定端口，0 表示随机端口）
    this.server = new AgentServer(config.agentPort || 0);
  }

  // ==================== 工厂方法 ====================

  /**
   * 创建 A2AAgent 实例
   * @param config A2A 配置
   */
  static create(config: A2AConfig): A2AAgent {
    return new A2AAgent(config);
  }

  // ==================== 生命周期 ====================

  /**
   * 启动 Agent
   * 1. 启动 gRPC 服务端
   * 2. 向注册中心注册（如果 autoRegister=true）
   */
  async start(): Promise<void> {
    if (this.started) {
      throw new Error('A2AAgent 已经启动');
    }

    try {
      // 1. 启动 gRPC 服务端
      await this.server.start();

      // 获取实际绑定的端口（当配置为 0 时）
      const actualPort = this.server.getPort();
      if (actualPort !== this.config.agentPort) {
        console.log(`[A2AAgent] 服务端实际端口: ${actualPort}`);
        // 更新配置中的端口，注册时使用实际端口
        (this.config as any).agentPort = actualPort;
      }

      // 2. 向注册中心注册
      if (this.config.autoRegister !== false) {
        await this.client.register();
      }

      this.started = true;
      console.log(
        `[A2AAgent] Agent 已启动: ${this.config.agentId} (port=${actualPort})`
      );
    } catch (err) {
      // 启动失败时清理资源
      this.server.stop();
      throw new Error(`A2AAgent 启动失败: ${formatError(err)}`);
    }
  }

  /**
   * 关闭 Agent
   * 1. 从注册中心注销
   * 2. 取消所有 PubSub 订阅
   * 3. 停止 gRPC 服务端
   */
  close(): void {
    // 取消所有订阅
    for (const subscriber of this.subscribers) {
      subscriber.unsubscribe();
    }
    this.subscribers = [];

    // 关闭客户端（注销 + 停止心跳）
    this.client.close();

    // 停止服务端
    this.server.stop();

    this.started = false;
    console.log(`[A2AAgent] Agent 已关闭: ${this.config.agentId}`);
  }

  // ==================== 处理器注册 ====================

  /**
   * 注册同步请求处理器
   * 当其他 Agent 调用 sendRequest 时触发
   * @param handler 请求处理函数
   */
  onRequest(handler: RequestHandler): this {
    this.server.onRequest(handler);
    return this;
  }

  /**
   * 注册单向消息处理器
   * 当其他 Agent 调用 send 时触发
   * @param handler 消息处理函数
   */
  onMessage(handler: MessageHandler): this {
    this.server.onMessage(handler);
    return this;
  }

  /**
   * 注册任务处理器
   * 当其他 Agent 调用 delegateTask 时触发
   * @param handler 任务处理函数
   */
  onTask(handler: TaskHandler): this {
    this.server.onTask(handler);
    return this;
  }

  /**
   * 订阅 PubSub topics
   * @param topics 要订阅的 topic 列表
   * @param handler 消息处理器
   */
  subscribe(topics: string[], handler: PubSubHandler): this {
    const subscriber = new PubSubSubscriber({
      registryHost: this.config.registryHost,
      registryPort: this.config.registryPort,
      agentId: this.config.agentId,
      jwtToken: this.config.jwtToken,
      sessionToken: this.client.getSessionToken(),
      connectTimeoutMs: this.config.connectTimeoutMs,
    });

    subscriber.subscribe(topics, handler);
    this.subscribers.push(subscriber);
    return this;
  }

  // ==================== 客户端方法代理 ====================

  /**
   * 按类型发现 Agent 列表
   * @param agentType Agent 类型
   */
  discover(agentType: string): Promise<AgentInfo[]> {
    return this.client.discover(agentType);
  }

  /**
   * 按能力发现 Agent 列表
   * @param capability 能力名称
   */
  discoverByCapability(capability: string): Promise<AgentInfo[]> {
    return this.client.discoverByCapability(capability);
  }

  /**
   * 获取指定 Agent 信息
   * @param agentId Agent ID
   */
  getAgent(agentId: string): Promise<AgentInfo | null> {
    return this.client.getAgent(agentId);
  }

  /**
   * 发送同步请求并等待响应
   * @param toAgentId 目标 Agent ID
   * @param payload 请求载荷
   * @param timeoutMs 超时毫秒数
   */
  sendRequest(
    toAgentId: string,
    payload: Record<string, any>,
    timeoutMs?: number
  ): Promise<Record<string, any>> {
    return this.client.sendRequest(toAgentId, payload, timeoutMs);
  }

  /**
   * 发送单向消息（fire-and-forget）
   * @param toAgentId 目标 Agent ID
   * @param topic 消息主题
   * @param payload 消息载荷
   */
  send(
    toAgentId: string,
    topic: string,
    payload: Record<string, any>
  ): Promise<string> {
    return this.client.send(toAgentId, topic, payload);
  }

  /**
   * 发布消息到 topic
   * @param topic 消息主题
   * @param payload 消息载荷
   */
  publish(topic: string, payload: Record<string, any>): Promise<string> {
    return this.client.publish(topic, payload);
  }

  /**
   * 委托任务给目标 Agent
   * @param toAgentId 目标 Agent ID
   * @param taskType 任务类型
   * @param input 任务输入
   * @param timeoutSeconds 任务超时（秒）
   */
  delegateTask(
    toAgentId: string,
    taskType: string,
    input: Record<string, any>,
    timeoutSeconds?: number
  ): Promise<string> {
    return this.client.delegateTask(toAgentId, taskType, input, timeoutSeconds);
  }

  /**
   * 查询任务状态
   * @param taskId 任务 ID
   */
  getTaskStatus(taskId: string): Promise<Task> {
    return this.client.getTaskStatus(taskId);
  }

  /**
   * 取消任务
   * @param taskId 任务 ID
   * @param reason 取消原因
   */
  cancelTask(taskId: string, reason?: string): Promise<boolean> {
    return this.client.cancelTask(taskId, reason);
  }

  // ==================== 状态查询 ====================

  /**
   * 是否已启动
   */
  isStarted(): boolean {
    return this.started;
  }

  /**
   * 获取 Agent ID
   */
  getAgentId(): string {
    return this.config.agentId;
  }

  /**
   * 获取服务端监听端口
   */
  getPort(): number {
    return this.server.getPort();
  }

  /**
   * 获取底层 A2AClient（用于高级操作）
   */
  getClient(): A2AClient {
    return this.client;
  }

  /**
   * 获取底层 AgentServer（用于高级操作）
   */
  getServer(): AgentServer {
    return this.server;
  }
}
